### PR TITLE
Fixing possible overflow during multiplication

### LINKED ISSUE #1380: arithmetic operation overflow on ARM
I left `Parity/v1.2.0-unstable-69c29fc-20160620/arm-linux-gnu/rustc1.11.0-dev` running but later found out it'd crashed (I guess it must have been close to fully synced at the time). 

Not sure if it's a random crash related to an HDD connected via USB (there are huge amounts of DEBUG messages in the logs relating to DMA) or a real issue. My guess is the world may have stopped due to lots of SD card `syslog` writes which in turn could have starved parity.

```shell
#1747187 4618…f00f      1 blk/s    5 tx/s   0 Mgas/s    1/ 7 peers   #1747190     0+    0 Qed     46 MiB db   11 MiB chain    2 KiB queue   36 KiB sync

thread 'IO Worker #2' panicked at 'arithmetic operation overflow', util/bigint/src/uint.rs:1137
note: Run with `RUST_BACKTRACE=1` for a backtrace.
Aborted
````

Restarting parity works fine again without any corruptions, getting to fully synced again.

### PR DESCRIPTION
**Pull request by tomusdrw**
It was haunting me since I finished working on uint. In a following code there can be only one overflow, but cannot be sure when it will occur.

Most probably closes #1380

### CODE REVIEW

File: util/bigint/src/uint.rs [hunk 1]
```diff
@@ -425,12 +425,16 @@ macro_rules! uint_overflowing_mul_reg {
...
+				// No overflow here
+				let res = (c_u >> 32) + (overflow_u << 32);
+				// possible overflows
+				let (res, o1) = res.overflowing_add(overflow_l);
+				let (res, o2) = res.overflowing_add(carry2);
+				let (res, o3) = res.overflowing_add(ret[i + j + 1]);
+				ret[i + j + 1] = res;
+
+				// Only single overflow possible there
+				carry2 = (o1 | o2 | o3) as u64;
```
Comment (NikVolf) [inline util/bigint/src/uint.rs]: why only single?

File: util/bigint/src/uint.rs [hunk 2]
```diff
@@ -1305,12 +1309,16 @@ impl U256 {
...
+				// possible overflows
+				let (res, o1) = res.overflowing_add(overflow_l);
+				let (res, o2) = res.overflowing_add(carry2);
+				let (res, o3) = res.overflowing_add(ret[i + j + 1]);
+				ret[i + j + 1] = res;
 
-				ret[i + j + 1] = carry;
-				carry2 = o as u64;
+				// Only single overflow possible there
+				carry2 = (o1 | o2 | o3) as u64;
```
Comment (NikVolf) [inline util/bigint/src/uint.rs]: why only single ?

File: util/bigint/src/uint.rs (hunk 1 above)
Comment (tomusdrw) [inline util/bigint/src/uint.rs]: We are multiplying two 64bit numbers - the result will fit in 128bits.
Those operations are adding the possible carrry (so two 64bits, split into couple of parts) - the result will fit in 64bits + 1 carry.

Comment (NikVolf) [inline util/bigint/src/uint.rs]: but you then add result to the next word, don't you ?..
`let (res, o3) = res.overflowing_add(ret[i + j + 1]);`
which can be also overflow

Comment (NikVolf) [inline util/bigint/src/uint.rs]: though forget it, i think overflow tests are made for each word there

**Pull request merged**